controller_manager/src/controller_manager.cpp

```diff
@@ -53,61 +58,787 @@ ControllerManager::load_controller(
+    to.push_back(from_controller);
+  }
+
+  // Checks that we're not duplicating controllers
+  for (const auto & to_controller : to) {
+    if (controller.info.name == to_controller.info.name) {
+      to.clear();
+      RCLCPP_ERROR(
+        get_logger(),
+        "A controller named '%s' was already loaded inside the controller manager",
+        controller.info.name.c_str());
+      return nullptr;
+    }
+  }
+
+  controller.c->init(hw_, controller.info.name);
+
+  // TODO(v-lopez) this should only be done if controller_manager is configured.
+  // Probably the whole load_controller part should fail if the controller_manager
+  // is not configured, should it implement a LifecycleNodeInterface
+  // https://github.com/ros-controls/ros2_control/issues/152
+  controller.c->get_lifecycle_node()->configure();
+  executor_->add_node(controller.c->get_lifecycle_node()->get_node_base_interface());
+  to.emplace_back(controller);
+
+  // Destroys the old controllers list when the realtime thread is finished with it.
+  RCLCPP_DEBUG(get_logger(), "Realtime switches over to new controller list");
+  rt_controllers_wrapper_.switch_updated_list(guard);
+  RCLCPP_DEBUG(get_logger(), "Destruct controller");
+  std::vector<ControllerSpec> & new_unused_list = rt_controllers_wrapper_.get_unused_list(
+    guard);
+  new_unused_list.clear();
+  RCLCPP_DEBUG(get_logger(), "Destruct controller finished");
+
+  return to.back().c;
+}
+
+controller_interface::ControllerInterfaceSharedPtr ControllerManager::get_controller_by_name(
+  const std::string & name)
+{
+  // lock controllers
+  std::lock_guard<std::recursive_mutex> guard(rt_controllers_wrapper_.controllers_lock_);
+
+  for (const auto & controller : rt_controllers_wrapper_.get_updated_list(guard)) {
+    if (controller.info.name == name) {
+      return controller.c;
+    }
+  }
+  return nullptr;
+}
+
+void ControllerManager::manage_switch()
+{
+#ifdef TODO_IMPLEMENT_RESOURCE_CHECKING
+  // switch hardware interfaces (if any)
+  if (!switch_params_.started) {
+    robot_hw_->doSwitch(switch_start_list_, switch_stop_list_);
+    switch_params_.started = true;
+  }
+#endif
+
+  stop_controllers();
+
+  // start controllers once the switch is fully complete
+  if (!switch_params_.start_asap) {
+    start_controllers();
+  } else {
+    // start controllers as soon as their required joints are done switching
+    start_controllers_asap();
+  }
+}
+
+void ControllerManager::stop_controllers()
+{
+  // stop controllers
+  for (const auto & request : stop_request_) {
+    if (is_controller_running(*request)) {
+      const auto new_state = request->get_lifecycle_node()->deactivate();
+      if (new_state.id() != lifecycle_msgs::msg::State::PRIMARY_STATE_INACTIVE) {
+        RCLCPP_ERROR(
+          get_logger(),
+          "After deactivating, controller %s is in state %s, expected Inactive",
+          request->get_lifecycle_node()->get_name(),
+          new_state.label().c_str());
+      }
+    }
+  }
```

Comment:
`std::find_if`?

Reply:
Done

Reply:
is it?

Reply:
There was nothing to be `find_if`ed here, it was the complete list of controllers that needed to be stopped.